metals/src/main/scala/scala/meta/internal/metals/ScalafixProvider.scala

```diff
@@ -35,52 +39,76 @@ case class ScalafixProvider(
       scribe.info(s"Organize imports requires saving the file first")
       languageClient.showMessage(
         MessageType.Warning,
-        s"Save ${file.toNIO.getFileName} before organizing imports"
+        s"Save ${file.toNIO.getFileName} to compile it before organizing imports"
       )
-      Nil
+      Future.successful(Nil)
     } else {
-      val scalafixConfPath = workspace.resolve(scalafixConfigPath)
-      val scalafixConf =
-        if (scalafixConfPath.isFile) Some(scalafixConfPath.toNIO)
-        else None
-      val scalafixEvaluation = for {
-        api <- getScalafix(scalaBinaryVersion)
-        urlClassLoaderWithExternalRule <- getRuleClassLoader(
-          scalaBinaryVersion,
-          api.getClass.getClassLoader
-        )
-      } yield {
-        val scalacOption =
-          if (scalaBinaryVersion == "2.13") "-Wunused:imports"
-          else "-Ywarn-unused-import"
-
-        api
-          .newArguments()
-          .withScalaVersion(scalaVersion)
-          .withClasspath(classpath)
-          .withToolClasspath(urlClassLoaderWithExternalRule)
-          .withConfig(scalafixConf.asJava)
-          .withRules(List(organizeImportRuleName).asJava)
-          .withPaths(List(file.toNIO).asJava)
-          .withSourceroot(workspace.toNIO)
-          .withScalacOptions(Collections.singletonList(scalacOption))
-          .evaluate()
-      }
+      compilations.compilationFinished(file).map { _ =>
+        val scalafixConfPath = workspace.resolve(scalafixConfigPath)
+        val scalafixConf =
+          if (scalafixConfPath.isFile) Some(scalafixConfPath.toNIO)
+          else None
+        val scalafixEvaluation = for {
+          api <- getScalafix(scalaBinaryVersion)
+          urlClassLoaderWithExternalRule <- getRuleClassLoader(
+            scalaBinaryVersion,
+            api.getClass.getClassLoader
+          )
+        } yield {
+          val scalacOption =
+            if (scalaBinaryVersion == "2.13") "-Wunused:imports"
+            else "-Ywarn-unused-import"
+
+          api
+            .newArguments()
+            .withScalaVersion(scalaVersion)
+            .withClasspath(classpath)
+            .withToolClasspath(urlClassLoaderWithExternalRule)
+            .withConfig(scalafixConf.asJava)
+            .withRules(List(organizeImportRuleName).asJava)
+            .withPaths(List(file.toNIO).asJava)
+            .withSourceroot(workspace.toNIO)
+            .withScalacOptions(Collections.singletonList(scalacOption))
+            .evaluate()
+        }
+        val edits = for {
+          results <- scalafixEvaluation
+          if results.isSuccessful()
+          fileEvaluation <- results.getFileEvaluations().headOption
+          patches <- fileEvaluation.previewPatches().asScala
+        } yield textEditsFrom(patches, fileInput)
+
+        if (!scalafixEvaluation.exists(_.isSuccessful())) {
+          val scalafixError =
+            scalafixEvaluation.flatMap(_.getMessageError().asScala)
```

Comment:
You are right, getMessageError, is empty except if the arguments for scalafix are missing or some other parsing error.